src/gc/gc.cpp

```diff
@@ -36101,43 +36101,21 @@ CFinalize::FinalizeSegForAppDomain (AppDomain *pDomain,
         // if it has the index we are looking for. If the methodtable is null, it can't be from the
         // unloading domain, so skip it.
         if (method_table(obj) == NULL)
+        {
             continue;
+        }
 
-        // eagerly finalize all objects except those that may be agile.
+        // skip objects not in the unloading appdomain
         if (obj->GetAppDomainIndex() != pDomain->GetIndex())
```

Comment:
This should be inside the ShouldFinalizeObjectForUnload callback as well. `obj->GetAppDomainIndex()` and `pDomain->GetIndex()` are bad dependencies.